pyrokinetics/pyro.py

```diff
@@ -1,339 +1,1423 @@
+                warn_msg = (
+                    f"Provided template file '{template_file}' to write_gk_file, but "
+                    "there is already data available for the gyrokinetics code "
+                    f"{gk_code}. Ignoring the template file."
+                )
+                warnings.warn(warn_msg, UserWarning)
+            # TODO Should this overwrite with the current context? It'll do so if the
+            # new context doesn't already exist, but if it does already exist it won't
+            # be changed.
+            self._switch_gk_context(
+                gk_code, template_file=template_file, force_overwrite=False
+            )
 
-        # if self.gk_code is still None, we couldn't infer a file type
-        if self.gk_code is None:
-            raise ValueError("Could not determine gk_code from file type")
+        # Update to account for any changes to this context
+        self.update_gk_code()
 
-        # If equilibrium already loaded then it won't load the input file
-        template = hasattr(self, "eq")
+        # Set file info in new context
+        self.gk_file = Path(file_name)
 
-        # Better way to select code?
-        self.gk_code.read(self, self.gk_file, template)
+        # Write to disk
+        self.gk_input.write(self.gk_file, float_format=self.float_format)
 
-    def write_gk_file(self, file_name, template_file=None, directory=".", gk_code=None):
+        # Switch back to original context
+        self._switch_gk_context(prev_gk_code, force_overwrite=False)
+
+    def load_gk_output(self, path: Optional[PathLike] = None, **kwargs) -> None:
         """
-        Writes single GK input file to file_name
+        Loads gyrokinetics output into Xarray Dataset.
+
+        Sets the attributes gk_output and gk_output_file for the current context. If
+        provided with a path, will attempt to read output from that path. Otherwise,
+        will infer the path from the value of ``gk_file``.
+
+        Parameters
+        ----------
+        path: PathLike, default None
+            Path to the gyrokinetics output file/directory. Valid ``path`` for each code
+            are:
+
+            - GS2: Path to '\\*.out.nc' NetCDF4 file
+            - CGYRO: Path to directory containing output files
+            - GENE: Path to directory containing output files if numbered 0000,\
+            otherwise provide one filename from parameters_####, nrg_#### or field_####.\
+            Pyrokinetics will search for the other files in the same directory.
 
+            If set to None, infers path from ``gk_file``.
+        **kwargs
+            Arguments to pass to the ``GKOutputReader``.
 
+        Returns
+        -------
+        ``None``
+
+        Raises
+        ------
+        RuntimeError
+            If no path is provided, and no ``gk_file`` exists.
+        Exception
+            Various errors may occur while processing a gyrokinetics output.
         """
+        if path is None:
```

Comment:
Is it possible to move all this logic into the respective `GKInput` readers? This current method is not very extensible. They already have some related handling. Do they need a (static) method to return the appropriate `gk_output_file`?

Reply:
I've moved the logic into the `GKOutputReaders`. The way in which `GKOutputReaders` are accessed from `Pyro` feels a bit clunky though, so it might be a good idea to return to this later.